`stack script` ignores `nix:  enable: true` in `~/.stack/config.yaml`

I want to make scripts that work on both nix and non-nix OS without changes

### Steps to reproduce

example-script.hs
```hs
#!/usr/bin/env stack
{- stack
  script
  --resolver lts-9.18
-}

import           System.Environment    (getArgs)

main :: IO ()
main = do
  args <- getArgs
  print args
```

~/.stack/config.yaml
```
nix:
  enable: true
  packages:
    - libcxx
    - icu
    - gcc
    - ncurses
    - zlib
```

```sh
 ~/projects/nuuz/nuuz-chrome-extension/packages/metalib   master  update-index-js.hs index.js
Already downloaded.
Running /home/bjorn/.stack/programs/x86_64-linux/ghc-8.0.2.temp/ghc-8.0.2/configure --prefix=/home/bjorn/.stack/programs/x86_64-linux/ghc-8.0.2/ in directory /home/bjorn/.stack/programs/x86_64-linux/ghc-8.0.2.temp/ghc-8.0.2/ exited with ExitFailure 1

checking for path to top of build tree...
/home/bjorn/.stack/programs/x86_64-linux/ghc-8.0.2.temp/ghc-8.0.2/configure: utils/ghc-pwd/dist-install/build/tmp/ghc-pwd-bindist: /bin/bash: bad interpreter: No such file or directory
configure: error: cannot determine current directory


Error: Error encountered while configuring GHC with
         /home/bjorn/.stack/programs/x86_64-linux/ghc-8.0.2.temp/ghc-8.0.2/configure --prefix=/home/bjorn/.stack/programs/x86_64-linux/ghc-8.0.2/
         run in /home/bjorn/.stack/programs/x86_64-linux/ghc-8.0.2.temp/ghc-8.0.2/

The following directories may now contain files, but won't be used by stack:
  - /home/bjorn/.stack/programs/x86_64-linux/ghc-8.0.2.temp/
  - /home/bjorn/.stack/programs/x86_64-linux/ghc-8.0.2/

Configuring GHC ...% 
```

But I can resolve this issue by adding `--nix` option
```hs
#!/usr/bin/env stack
{- stack
  script
  --nix
  --resolver lts-9.18
-}
```

### Expected
To work without `--nix`

### Stack version
```
 ✘  ~  stack --version
Version 1.5.1 x86_64
```

Sorry, can't upgrade to latest stack-1.6.3

### Method of installation

* nix

P.S. [without --nix, but with --verbose](https://pastebin.com/rbNnrjZe)

--- a/.azure/azure-nightly-template-linux.yml
+++ b/.azure/azure-nightly-template-linux.yml
@@ -37,7 +37,7 @@ jobs:
       # FIXME: Note that cabal-install might not be required once
       # https://github.com/commercialhaskell/stack/issues/4410 get's
       # fixed.
-      stack test --flag stack:integration-tests stack:test:stack-integration-test --interleaved-output
+      stack build --flag stack:integration-tests stack --interleaved-output --exec stack-integration-test
       set +ex
     displayName: Integration Test
   - script: |

--- a/.azure/azure-nightly-template-linux.yml
+++ b/.azure/azure-nightly-template-linux.yml
@@ -37,7 +37,7 @@ jobs:
       # FIXME: Note that cabal-install might not be required once
       # https://github.com/commercialhaskell/stack/issues/4410 get's
       # fixed.
-      stack build --flag stack:integration-tests stack --interleaved-output --exec stack-integration-test
+      etc/scripts/integration-tests.sh
       set +ex
     displayName: Integration Test
   - script: |

--- a/.azure/azure-nightly-template-osx.yml
+++ b/.azure/azure-nightly-template-osx.yml
@@ -39,7 +39,7 @@ jobs:
       # FIXME: Note that cabal-install might not be required once
       # https://github.com/commercialhaskell/stack/issues/4410 get's
       # fixed.
-      stack test --flag stack:integration-tests stack:test:stack-integration-test --interleaved-output
+      stack build --flag stack:integration-tests stack --interleaved-output --exec stack-integration-test
       set +ex
     displayName: Integration Test
   - script: |

--- a/.azure/azure-nightly-template-osx.yml
+++ b/.azure/azure-nightly-template-osx.yml
@@ -39,7 +39,7 @@ jobs:
       # FIXME: Note that cabal-install might not be required once
       # https://github.com/commercialhaskell/stack/issues/4410 get's
       # fixed.
-      stack build --flag stack:integration-tests stack --interleaved-output --exec stack-integration-test
+      etc/scripts/integration-tests.sh
       set +ex
     displayName: Integration Test
   - script: |

--- a/.azure/azure-nightly-template-windows.yml
+++ b/.azure/azure-nightly-template-windows.yml
@@ -43,7 +43,7 @@ jobs:
       # FIXME: Note that cabal-install might not be required once
       # https://github.com/commercialhaskell/stack/issues/4410 get's
       # fixed.
-      stack build --flag stack:integration-tests stack --interleaved-output --exec stack-integration-test
+      etc/scripts/integration-tests.sh
       set +ex
     displayName: Integration Test
   - powershell: |

--- a/.azure/azure-nightly-template-windows.yml
+++ b/.azure/azure-nightly-template-windows.yml
@@ -43,7 +43,7 @@ jobs:
       # FIXME: Note that cabal-install might not be required once
       # https://github.com/commercialhaskell/stack/issues/4410 get's
       # fixed.
-      stack test --flag stack:integration-tests stack:test:stack-integration-test --interleaved-output --no-terminal
+      stack build --flag stack:integration-tests stack --interleaved-output --exec stack-integration-test
       set +ex
     displayName: Integration Test
   - powershell: |

--- a/ChangeLog.md
+++ b/ChangeLog.md
@@ -178,6 +178,9 @@ Other enhancements:
   packages. See [#2465](https://github.com/commercialhaskell/stack/issues/2465)
 * Store caches in SQLite database instead of files.
 * No longer use "global" Docker image database (`docker.db`).
+* User config files are respected for the script command. See
+  [#3705](https://github.com/commercialhaskell/stack/issues/3705),
+  [#3887](https://github.com/commercialhaskell/stack/issues/3887).
 
 Bug fixes:
 

--- a/etc/scripts/integration-tests.sh
+++ b/etc/scripts/integration-tests.sh
@@ -1,0 +1,3 @@
+#!/bin/sh
+
+exec stack build --flag stack:integration-tests stack --interleaved-output --exec stack-integration-test

--- a/package.yaml
+++ b/package.yaml
@@ -277,17 +277,6 @@ executables:
       cpp-options: -DHIDE_DEP_VERSIONS
     - condition: flag(supported-build)
       cpp-options: -DSUPPORTED_BUILD
-tests:
-  stack-test:
-    main: Spec.hs
-    source-dirs: src/test
-    ghc-options:
-    - -threaded
-    dependencies:
-    - QuickCheck
-    - hspec
-    - stack
-    - smallcheck
   stack-integration-test:
     main: IntegrationSpec.hs
     source-dirs:
@@ -303,6 +292,17 @@ tests:
     when:
     - condition: ! '!(flag(integration-tests))'
       buildable: false
+tests:
+  stack-test:
+    main: Spec.hs
+    source-dirs: src/test
+    ghc-options:
+    - -threaded
+    dependencies:
+    - QuickCheck
+    - hspec
+    - stack
+    - smallcheck
 flags:
   static:
     description: Pass -static/-pthread to ghc when linking the stack binary.

--- a/src/Stack/Build/Target.hs
+++ b/src/Stack/Build/Target.hs
@@ -479,4 +479,4 @@ parseTargets needTargets haddockDeps boptscli smActual = do
       case configProject $ bcConfig bconfig of
         PCProject _ -> False
         PCGlobalProject -> True
-        PCNoConfig _ -> False
+        PCNoProject _ -> False

--- a/src/Stack/Build/Target.hs
+++ b/src/Stack/Build/Target.hs
@@ -478,5 +478,5 @@ parseTargets needTargets haddockDeps boptscli smActual = do
     bcImplicitGlobal bconfig =
       case configProject $ bcConfig bconfig of
         PCProject _ -> False
-        PCNoProject -> True
+        PCGlobalProject -> True
         PCNoConfig _ -> False

--- a/src/Stack/Config.hs
+++ b/src/Stack/Config.hs
@@ -355,8 +355,16 @@ configFromConfigMonoid
          Nothing -> pure defaultHackageSecurityConfig
          Just [hsc] -> pure hsc
          Just x -> error $ "When overriding the default package index, you must provide exactly one value, received: " ++ show x
+     mpantryRoot <- liftIO $ lookupEnv "PANTRY_ROOT"
+     pantryRoot <-
+       case mpantryRoot of
+         Just dir ->
+           case parseAbsDir dir of
+             Nothing -> throwString $ "Failed to parse PANTRY_ROOT environment variable (expected absolute directory): " ++ show dir
+             Just x -> pure x
+         Nothing -> pure $ configStackRoot </> relDirPantry
      withPantryConfig
-       (configStackRoot </> relDirPantry)
+       pantryRoot
        hsc
        (maybe HpackBundled HpackCommand $ getFirst configMonoidOverrideHpack)
        clConnectionCount

--- a/src/Stack/Config.hs
+++ b/src/Stack/Config.hs
@@ -216,7 +216,7 @@ configFromConfigMonoid
      let mproject =
            case configProject of
              PCProject pair -> Just pair
-             PCNoProject -> Nothing
+             PCGlobalProject -> Nothing
              PCNoConfig _deps -> Nothing
      configWorkDir0 <- maybe (return relDirStackWork) (liftIO . parseRelDir) mstackWorkEnv
      let configWorkDir = fromFirst configWorkDir0 configMonoidWorkDir
@@ -410,7 +410,7 @@ loadConfig inner = do
     let (mproject', addConfigMonoid) =
           case mproject of
             PCProject (proj, fp, cm) -> (PCProject (proj, fp), (cm:))
-            PCNoProject -> (PCNoProject, id)
+            PCGlobalProject -> (PCGlobalProject, id)
             PCNoConfig deps -> (PCNoConfig deps, id)
     let loadHelper inner2 = do
           userConfigPath <- getDefaultUserConfigPath stackRoot
@@ -435,7 +435,7 @@ loadConfig inner = do
     let withConfig = case mproject of
           PCNoConfig extraDeps -> configNoLocalConfig stackRoot mresolver configArgs extraDeps
           PCProject _project -> loadHelper
-          PCNoProject -> loadHelper
+          PCGlobalProject -> loadHelper
 
     withConfig $ \config -> do
       unless (mkVersion' Meta.version `withinRange` configRequireStackVersion config)
@@ -472,7 +472,7 @@ loadBuildConfig = do
       PCNoConfig extraDeps -> do
           p <- assert (isJust mresolver) (getEmptyProject mresolver extraDeps)
           return (p, configUserConfigPath config)
-      PCNoProject -> do
+      PCGlobalProject -> do
             logDebug "Run from outside a project, using implicit global project config"
             destDir <- getImplicitGlobalProjectDir config
             let dest :: Path Abs File
@@ -760,7 +760,7 @@ getProjectConfig :: HasLogFunc env
                  -- ^ Override stack.yaml
                  -> RIO env (ProjectConfig (Path Abs File))
 getProjectConfig (SYLOverride stackYaml) = return $ PCProject stackYaml
-getProjectConfig SYLNoProject = return PCNoProject
+getProjectConfig SYLGlobalProject = return PCGlobalProject
 getProjectConfig SYLDefault = do
     env <- liftIO getEnvironment
     case lookup "STACK_YAML" env of
@@ -769,7 +769,7 @@ getProjectConfig SYLDefault = do
             liftM PCProject $ resolveFile' fp
         Nothing -> do
             currDir <- getCurrentDir
-            maybe PCNoProject PCProject <$> findInParents getStackDotYaml currDir
+            maybe PCGlobalProject PCProject <$> findInParents getStackDotYaml currDir
   where
     getStackDotYaml dir = do
         let fp = dir </> stackDotYaml
@@ -796,9 +796,9 @@ loadProjectConfig mstackYaml = do
             logDebug $ "Loading project config file " <>
                         fromString (maybe (toFilePath fp) toFilePath (stripProperPrefix currDir fp))
             PCProject <$> load fp
-        PCNoProject -> do
+        PCGlobalProject -> do
             logDebug "No project config file found, using defaults."
-            return PCNoProject
+            return PCGlobalProject
         PCNoConfig extraDeps -> do
             logDebug "Ignoring config files"
             return $ PCNoConfig extraDeps

--- a/src/Stack/Config.hs
+++ b/src/Stack/Config.hs
@@ -36,7 +36,6 @@ import           Control.Monad.Extra (firstJustM)
 import           Stack.Prelude
 import           Data.Aeson.Extended
 import qualified Data.ByteString as S
-import           Data.ByteString.Builder (toLazyByteString)
 import           Data.Coerce (coerce)
 import qualified Data.IntMap as IntMap
 import qualified Data.Map as Map
@@ -53,7 +52,6 @@ import           GHC.Conc (getNumProcessors)
 import           Lens.Micro ((.~))
 import           Network.HTTP.StackClient (httpJSON, parseUrlThrow, getResponseBody)
 import           Options.Applicative (Parser, strOption, long, help)
-import qualified Pantry.SHA256 as SHA256
 import           Path
 import           Path.Extra (toFilePathNoTrailingSep)
 import           Path.Find (findInParents)
@@ -174,41 +172,18 @@ getLatestResolver = do
                listToMaybe (reverse (IntMap.toList (snapshotsLts snapshots)))
     pure $ fromMaybe (nightlySnapshotLocation (snapshotsNightly snapshots)) mlts
 
--- | Create a 'Config' value when we're not using any local
--- configuration files (e.g., the script command)
-configNoLocalConfig
-    :: HasRunner env
-    => Path Abs Dir -- ^ stack root
-    -> Maybe AbstractResolver
-    -> ConfigMonoid
-    -> [PackageIdentifierRevision]
-    -> (Config -> RIO env a)
-    -> RIO env a
-configNoLocalConfig _ Nothing _ _ _ = throwIO NoResolverWhenUsingNoLocalConfig
-configNoLocalConfig stackRoot (Just resolver) configMonoid extraDeps inner = do
-    userConfigPath <- liftIO $ getFakeConfigPath stackRoot resolver
-    configFromConfigMonoid
-      stackRoot
-      userConfigPath
-      False
-      (Just resolver)
-      (PCNoConfig extraDeps)
-      configMonoid
-      inner
-
 -- Interprets ConfigMonoid options.
 configFromConfigMonoid
     :: HasRunner env
     => Path Abs Dir -- ^ stack root, e.g. ~/.stack
     -> Path Abs File -- ^ user config file path, e.g. ~/.stack/config.yaml
-    -> Bool -- ^ allow locals?
     -> Maybe AbstractResolver
     -> ProjectConfig (Project, Path Abs File)
     -> ConfigMonoid
     -> (Config -> RIO env a)
     -> RIO env a
 configFromConfigMonoid
-    configStackRoot configUserConfigPath configAllowLocals configResolver
+    configStackRoot configUserConfigPath configResolver
     configProject ConfigMonoid{..} inner = do
      -- If --stack-work is passed, prefer it. Otherwise, if STACK_WORK
      -- is set, use that. If neither, use the default ".stack-work"
@@ -217,7 +192,12 @@ configFromConfigMonoid
            case configProject of
              PCProject pair -> Just pair
              PCGlobalProject -> Nothing
-             PCNoConfig _deps -> Nothing
+             PCNoProject _deps -> Nothing
+         configAllowLocals =
+           case configProject of
+             PCProject _ -> True
+             PCGlobalProject -> True
+             PCNoProject _ -> False
      configWorkDir0 <- maybe (return relDirStackWork) (liftIO . parseRelDir) mstackWorkEnv
      let configWorkDir = fromFirst configWorkDir0 configMonoidWorkDir
          configLatestSnapshot = fromFirst
@@ -411,31 +391,25 @@ loadConfig inner = do
           case mproject of
             PCProject (proj, fp, cm) -> (PCProject (proj, fp), (cm:))
             PCGlobalProject -> (PCGlobalProject, id)
-            PCNoConfig deps -> (PCNoConfig deps, id)
-    let loadHelper inner2 = do
-          userConfigPath <- getDefaultUserConfigPath stackRoot
-          extraConfigs0 <- getExtraConfigs userConfigPath >>=
-              mapM (\file -> loadConfigYaml (parseConfigMonoid (parent file)) file)
-          let extraConfigs =
-                -- non-project config files' existence of a docker section should never default docker
-                -- to enabled, so make it look like they didn't exist
-                map (\c -> c {configMonoidDockerOpts =
-                                  (configMonoidDockerOpts c) {dockerMonoidDefaultEnable = Any False}})
-                    extraConfigs0
+            PCNoProject deps -> (PCNoProject deps, id)
 
+    userConfigPath <- getDefaultUserConfigPath stackRoot
+    extraConfigs0 <- getExtraConfigs userConfigPath >>=
+        mapM (\file -> loadConfigYaml (parseConfigMonoid (parent file)) file)
+    let extraConfigs =
+          -- non-project config files' existence of a docker section should never default docker
+          -- to enabled, so make it look like they didn't exist
+          map (\c -> c {configMonoidDockerOpts =
+                            (configMonoidDockerOpts c) {dockerMonoidDefaultEnable = Any False}})
+              extraConfigs0
+
+    let withConfig =
           configFromConfigMonoid
             stackRoot
             userConfigPath
-            True -- allow locals
             mresolver
             mproject'
             (mconcat $ configArgs : addConfigMonoid extraConfigs)
-            inner2
-
-    let withConfig = case mproject of
-          PCNoConfig extraDeps -> configNoLocalConfig stackRoot mresolver configArgs extraDeps
-          PCProject _project -> loadHelper
-          PCGlobalProject -> loadHelper
 
     withConfig $ \config -> do
       unless (mkVersion' Meta.version `withinRange` configRequireStackVersion config)
@@ -469,8 +443,11 @@ loadBuildConfig = do
       PCProject (project, fp) -> do
           forM_ (projectUserMsg project) (logWarn . fromString)
           return (project, fp)
-      PCNoConfig extraDeps -> do
-          p <- assert (isJust mresolver) (getEmptyProject mresolver extraDeps)
+      PCNoProject extraDeps -> do
+          p <-
+            case mresolver of
+              Nothing -> throwIO NoResolverWhenUsingNoProject
+              Just _ -> getEmptyProject mresolver extraDeps
           return (p, configUserConfigPath config)
       PCGlobalProject -> do
             logDebug "Run from outside a project, using implicit global project config"
@@ -779,7 +756,7 @@ getProjectConfig SYLDefault = do
         if exists
             then return $ Just fp
             else return Nothing
-getProjectConfig (SYLNoConfig extraDeps) = return $ PCNoConfig extraDeps
+getProjectConfig (SYLNoProject extraDeps) = return $ PCNoProject extraDeps
 
 -- | Find the project config file location, respecting environment variables
 -- and otherwise traversing parents. If no config is found, we supply a default
@@ -799,9 +776,9 @@ loadProjectConfig mstackYaml = do
         PCGlobalProject -> do
             logDebug "No project config file found, using defaults."
             return PCGlobalProject
-        PCNoConfig extraDeps -> do
+        PCNoProject extraDeps -> do
             logDebug "Ignoring config files"
-            return $ PCNoConfig extraDeps
+            return $ PCNoProject extraDeps
   where
     load fp = do
         iopc <- loadConfigYaml (parseProjectAndConfigMonoid (parent fp)) fp
@@ -842,29 +819,6 @@ getDefaultUserConfigPath stackRoot = do
         ensureDir (parent path)
         liftIO $ S.writeFile (toFilePath path) defaultConfigYaml
     return path
-
--- | Get a fake configuration file location, used when doing a "no
--- config" run (the script command).
-getFakeConfigPath
-    :: (MonadIO m, MonadThrow m)
-    => Path Abs Dir -- ^ stack root
-    -> AbstractResolver
-    -> m (Path Abs File)
-getFakeConfigPath stackRoot ar = do
-  asString <-
-    case ar of
-      ARResolver r -> pure $ T.unpack $ SHA256.toHexText $ SHA256.hashLazyBytes $ toLazyByteString $ getUtf8Builder $ display r
-      _ -> throwM $ InvalidResolverForNoLocalConfig $ show ar
-  -- This takeWhile is an ugly hack. We don't actually need this
-  -- path for anything useful. But if we take the raw value for
-  -- a custom snapshot, it will be unparseable in a PATH.
-  -- Therefore, we add in this silly "strip up to :".
-  -- Better would be to defer figuring out this value until
-  -- after we have a fully loaded snapshot with a hash.
-  asDir <- parseRelDir $ takeWhile (/= ':') asString
-  let full = stackRoot </> relDirScript </> asDir </> relFileConfigYaml
-  ensureDir (parent full)
-  return full
 
 packagesParser :: Parser [String]
 packagesParser = many (strOption (long "package" <> help "Additional packages that must be installed"))

--- a/src/Stack/ConfigCmd.hs
+++ b/src/Stack/ConfigCmd.hs
@@ -67,7 +67,7 @@ cfgCmdSet cmd = do
                      case mstackYaml of
                          PCProject stackYaml -> return stackYaml
                          PCGlobalProject -> liftM (</> stackDotYaml) (getImplicitGlobalProjectDir conf)
-                         PCNoConfig _extraDeps -> throwString "config command used when no local configuration available"
+                         PCNoProject _extraDeps -> throwString "config command used when no project configuration available" -- maybe modify the ~/.stack/config.yaml file instead?
                  CommandScopeGlobal -> return (configUserConfigPath conf)
     -- We don't need to worry about checking for a valid yaml here
     (config :: Yaml.Object) <-

--- a/src/Stack/ConfigCmd.hs
+++ b/src/Stack/ConfigCmd.hs
@@ -66,7 +66,7 @@ cfgCmdSet cmd = do
                      mstackYaml <- getProjectConfig mstackYamlOption
                      case mstackYaml of
                          PCProject stackYaml -> return stackYaml
-                         PCNoProject -> liftM (</> stackDotYaml) (getImplicitGlobalProjectDir conf)
+                         PCGlobalProject -> liftM (</> stackDotYaml) (getImplicitGlobalProjectDir conf)
                          PCNoConfig _extraDeps -> throwString "config command used when no local configuration available"
                  CommandScopeGlobal -> return (configUserConfigPath conf)
     -- We don't need to worry about checking for a valid yaml here

--- a/src/Stack/Freeze.hs
+++ b/src/Stack/Freeze.hs
@@ -26,7 +26,7 @@ freeze (FreezeOpts mode) = do
   let warn = logWarn "No project was found: nothing to freeze"
   case mproject of
     PCProject (p, _) -> doFreeze p mode
-    PCNoProject -> warn
+    PCGlobalProject -> warn
     PCNoConfig _ -> warn
 
 doFreeze ::

--- a/src/Stack/Freeze.hs
+++ b/src/Stack/Freeze.hs
@@ -27,7 +27,7 @@ freeze (FreezeOpts mode) = do
   case mproject of
     PCProject (p, _) -> doFreeze p mode
     PCGlobalProject -> warn
-    PCNoConfig _ -> warn
+    PCNoProject _ -> warn
 
 doFreeze ::
        (HasProcessContext env, HasLogFunc env, HasPantryConfig env)

--- a/src/Stack/Options/Completion.hs
+++ b/src/Stack/Options/Completion.hs
@@ -86,7 +86,7 @@ flagCompleter = buildConfigCompleter $ \input -> do
           case configProject (bcConfig bconfig) of
             PCProject (p, _) -> projectFlags p
             PCGlobalProject -> mempty
-            PCNoConfig _ -> mempty
+            PCNoProject _ -> mempty
         flagEnabled name fl =
             fromMaybe (C.flagDefault fl) $
             Map.lookup (C.flagName fl) $

--- a/src/Stack/Options/Completion.hs
+++ b/src/Stack/Options/Completion.hs
@@ -85,7 +85,7 @@ flagCompleter = buildConfigCompleter $ \input -> do
         prjFlags =
           case configProject (bcConfig bconfig) of
             PCProject (p, _) -> projectFlags p
-            PCNoProject -> mempty
+            PCGlobalProject -> mempty
             PCNoConfig _ -> mempty
         flagEnabled name fl =
             fromMaybe (C.flagDefault fl) $

--- a/src/Stack/Runners.hs
+++ b/src/Stack/Runners.hs
@@ -14,7 +14,7 @@ module Stack.Runners
     , withEnvConfig
     , withDefaultEnvConfig
     , withConfig
-    , withNoProject
+    , withGlobalProject
     , withRunnerGlobal
     , ShouldReexec (..)
     ) where
@@ -35,11 +35,11 @@ import           System.Console.ANSI (hSupportsANSIWithoutEmulation)
 import           System.Terminal (getTerminalWidth)
 
 -- | Ensure that no project settings are used when running 'withConfig'.
-withNoProject :: RIO Runner a -> RIO Runner a
-withNoProject inner = do
+withGlobalProject :: RIO Runner a -> RIO Runner a
+withGlobalProject inner = do
   oldSYL <- view stackYamlLocL
   case oldSYL of
-    SYLDefault -> local (set stackYamlLocL SYLNoProject) inner
+    SYLDefault -> local (set stackYamlLocL SYLGlobalProject) inner
     _ -> throwString "Cannot use this command with options which override the stack.yaml location"
 
 -- | Helper for 'withEnvConfig' which passes in some default arguments:

--- a/src/Stack/Script.hs
+++ b/src/Stack/Script.hs
@@ -68,7 +68,7 @@ scriptCmd opts = do
         fromString (toFilePath fp)
       SYLGlobalProject -> logError "Ignoring SYLGlobalProject for script command"
       SYLDefault -> return ()
-      SYLNoConfig _ -> assert False (return ())
+      SYLNoProject _ -> assert False (return ())
 
     file <- resolveFile' $ soFile opts
     let scriptDir = parent file
@@ -76,7 +76,7 @@ scriptCmd opts = do
             { globalConfigMonoid = (globalConfigMonoid go)
                 { configMonoidInstallGHC = FirstTrue $ Just True
                 }
-            , globalStackYaml = SYLNoConfig $ soScriptExtraDeps opts
+            , globalStackYaml = SYLNoProject $ soScriptExtraDeps opts
             }
 
     -- Optimization: if we're compiling, and the executable is newer

--- a/src/Stack/Script.hs
+++ b/src/Stack/Script.hs
@@ -66,7 +66,7 @@ scriptCmd opts = do
       SYLOverride fp -> logError $
         "Ignoring override stack.yaml file for script command: " <>
         fromString (toFilePath fp)
-      SYLNoProject -> logError "Ignoring SYLNoProject for script command"
+      SYLGlobalProject -> logError "Ignoring SYLGlobalProject for script command"
       SYLDefault -> return ()
       SYLNoConfig _ -> assert False (return ())
 

--- a/src/Stack/Types/Config.hs
+++ b/src/Stack/Types/Config.hs
@@ -364,7 +364,7 @@ configProjectRoot c =
   case configProject c of
     PCProject (_, fp) -> Just $ parent fp
     PCGlobalProject -> Nothing
-    PCNoConfig _deps -> Nothing
+    PCNoProject _deps -> Nothing
 
 -- | Which packages do ghc-options on the command line apply to?
 data ApplyGhcOptions = AGOTargets -- ^ all local targets
@@ -463,8 +463,9 @@ data StackYamlLoc
     -- ^ Use the standard parent-directory-checking logic
     | SYLOverride !(Path Abs File)
     -- ^ Use a specific stack.yaml file provided
-    | SYLNoConfig ![PackageIdentifierRevision]
-    -- ^ Extra dependencies included in the script command line.
+    | SYLNoProject ![PackageIdentifierRevision]
+    -- ^ Do not load up a project, just user configuration. Include
+    -- the given extra dependencies with the resolver.
     | SYLGlobalProject
     -- ^ Do not look for a project configuration, and use the implicit global.
     deriving Show
@@ -481,9 +482,8 @@ data ProjectConfig a
     | PCGlobalProject
     -- ^ No project was found when using 'SYLDefault'. Instead, use
     -- the implicit global.
-    | PCNoConfig ![PackageIdentifierRevision]
-    -- ^ Use a no config run, which explicitly ignores any local
-    -- configuration values. This comes from 'SYLNoConfig'.
+    | PCNoProject ![PackageIdentifierRevision]
+    -- ^ Use a no project run. This comes from 'SYLNoProject'.
 
 -- | Parsed global command-line options monoid.
 data GlobalOptsMonoid = GlobalOptsMonoid
@@ -1037,8 +1037,7 @@ data ConfigException
   | UserDoesn'tOwnDirectory (Path Abs Dir)
   | ManualGHCVariantSettingsAreIncompatibleWithSystemGHC
   | NixRequiresSystemGhc
-  | NoResolverWhenUsingNoLocalConfig
-  | InvalidResolverForNoLocalConfig String
+  | NoResolverWhenUsingNoProject
   | DuplicateLocalPackageNames ![(PackageName, [PackageLocation])]
   deriving Typeable
 instance Show ConfigException where
@@ -1142,8 +1141,7 @@ instance Show ConfigException where
         , configMonoidSystemGHCName
         , "' or disable the Nix integration."
         ]
-    show NoResolverWhenUsingNoLocalConfig = "When using the script command, you must provide a resolver argument"
-    show (InvalidResolverForNoLocalConfig ar) = "The script command requires a specific resolver, you provided " ++ ar
+    show NoResolverWhenUsingNoProject = "When using the script command, you must provide a resolver argument"
     show (DuplicateLocalPackageNames pairs) = concat
         $ "The same package name is used in multiple local packages\n"
         : map go pairs

--- a/src/Stack/Types/Config.hs
+++ b/src/Stack/Types/Config.hs
@@ -363,7 +363,7 @@ configProjectRoot :: Config -> Maybe (Path Abs Dir)
 configProjectRoot c =
   case configProject c of
     PCProject (_, fp) -> Just $ parent fp
-    PCNoProject -> Nothing
+    PCGlobalProject -> Nothing
     PCNoConfig _deps -> Nothing
 
 -- | Which packages do ghc-options on the command line apply to?
@@ -465,7 +465,7 @@ data StackYamlLoc
     -- ^ Use a specific stack.yaml file provided
     | SYLNoConfig ![PackageIdentifierRevision]
     -- ^ Extra dependencies included in the script command line.
-    | SYLNoProject
+    | SYLGlobalProject
     -- ^ Do not look for a project configuration, and use the implicit global.
     deriving Show
 
@@ -478,7 +478,7 @@ data ProjectConfig a
     = PCProject a
     -- ^ Normal run: we want a project, and have one. This comes from
     -- either 'SYLDefault' or 'SYLOverride'.
-    | PCNoProject
+    | PCGlobalProject
     -- ^ No project was found when using 'SYLDefault'. Instead, use
     -- the implicit global.
     | PCNoConfig ![PackageIdentifierRevision]

--- a/src/main/Main.hs
+++ b/src/main/Main.hs
@@ -672,7 +672,7 @@ upgradeCmd upgradeOpts' = do
       logError "You cannot use the --resolver option with the upgrade command"
       liftIO exitFailure
     Nothing ->
-      withNoProject $
+      withGlobalProject $
       upgrade
 #ifdef USE_GIT_INFO
         (either (const Nothing) (Just . giHash) $$tGitInfoCwdTry)
@@ -898,12 +898,12 @@ initCmd :: InitOpts -> RIO Runner ()
 initCmd initOpts = do
     pwd <- getCurrentDir
     go <- view globalOptsL
-    withNoProject $ withConfig YesReexec (initProject pwd initOpts (globalResolver go))
+    withGlobalProject $ withConfig YesReexec (initProject pwd initOpts (globalResolver go))
 
 -- | Create a project directory structure and initialize the stack config.
 newCmd :: (NewOpts,InitOpts) -> RIO Runner ()
 newCmd (newOpts,initOpts) =
-    withNoProject $ withConfig YesReexec $ do
+    withGlobalProject $ withConfig YesReexec $ do
         dir <- new newOpts (forceOverwrite initOpts)
         exists <- doesFileExist $ dir </> stackDotYaml
         when (forceOverwrite initOpts || not exists) $ do

--- a/test/integration/IntegrationSpec.hs
+++ b/test/integration/IntegrationSpec.hs
@@ -1,133 +1,222 @@
-{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE NoImplicitPrelude   #-}
+{-# LANGUAGE OverloadedStrings   #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 
-import           Control.Applicative
-import           Control.Arrow
-import           Control.Concurrent.Async
-import           Control.Exception
-import           Control.Monad
-import           Control.Monad.IO.Class
-import           Control.Monad.Trans.Resource
-import qualified Data.ByteString.Lazy         as L
-import           Data.Char
-import           Data.Conduit
-import           Data.Conduit.Binary          (sinkLbs)
-import           Data.Conduit.Filesystem      (sourceDirectoryDeep)
-import qualified Data.Conduit.List            as CL
-import           Data.Conduit.Process
-import           Data.List                    (isSuffixOf, stripPrefix, sort)
-import qualified Data.Map                     as Map
-import           Data.Maybe                   (fromMaybe)
-import           Data.Text.Encoding.Error     (lenientDecode)
-import qualified Data.Text.Lazy               as TL
-import qualified Data.Text.Lazy.Encoding      as TL
-import           Data.Typeable
-import           Prelude -- Fix redundant import warnings
-import           System.Directory
-import           System.Environment
+import           Conduit
+import           Data.List                (stripPrefix)
+import           RIO
+import           RIO.Directory            hiding (findExecutable)
+import           RIO.FilePath
+import           RIO.List                 (partition)
+import qualified RIO.Map                  as Map
+import           RIO.Process
+import qualified RIO.Set                  as Set
+import qualified RIO.Text                 as T
+import           System.Environment       (getArgs, lookupEnv, getExecutablePath)
 import           System.Exit
-import           System.FilePath
-import           System.IO.Temp
+import           System.Info (os)
 import           System.PosixCompat.Files
-import           Test.Hspec
+
+-- This code does not use a test framework so that we get direct
+-- control of how the output is displayed.
 
 main :: IO ()
-main = do
-    srcDir  <- canonicalizePath ""
-    currDir <- canonicalizePath "test/integration"
+main = runSimpleApp $ do
+  logInfo "Initiating Stack integration test running"
 
-    let findExe name = do
-            mexe <- findExecutable name
-            case mexe of
-                Nothing -> error $ name ++ " not found on PATH"
-                Just exe -> return exe
-    runghc <- findExe "runghc"
-    stack <- findExe "stack"
+  args <- liftIO getArgs
+  fast <-
+    case args of
+      [] -> pure Slow
+      ["--fast"] -> pure Fast
+      _ -> error $ "Unknown arguments: " ++ show args
 
-    let testDir = currDir </> "tests"
-    tests <- getDirectoryContents testDir >>= filterM (hasTest testDir) . sort
+  results <- runApp fast $ do
+    logInfo "Running with the following environment"
+    proc "env" [] runProcess_
+    tests <- asks appTestDirs
+    let count = Set.size tests
+        loop !idx rest !accum =
+          case rest of
+            [] -> pure accum
+            next:rest' -> do
+              logInfo $ "Running integration test "
+                     <> display idx
+                     <> "/"
+                     <> display count
+                     <> ": "
+                     <> fromString (takeFileName next)
+              res <- test next
+              loop (idx + 1) rest' (res <> accum)
 
-    envOrig <- getEnvironment
+    loop (1 :: Int) (Set.toList tests) mempty
 
-    withSystemTempDirectory "stackhome" $ \newHome -> do
-        defaultStackRoot <- getAppUserDataDirectory "stack"
-        let newStackRoot = newHome </> takeFileName defaultStackRoot
-            env' = Map.toList
-                 $ Map.insert "STACK_EXE" stack
-                 $ Map.insert "HOME" newHome
-                 $ Map.insert "APPDATA" newHome
-                 $ Map.insert "STACK_ROOT" newStackRoot
-                 $ Map.insert "SRC_DIR" srcDir
-                 $ Map.delete "GHC_PACKAGE_PATH"
-                 $ Map.fromList
-                 $ map (first (map toUpper)) envOrig
-            origStackRoot = fromMaybe defaultStackRoot (lookup "STACK_ROOT" envOrig)
+  let (successes, failures) = partition ((== ExitSuccess) . snd)
+                            $ Map.toList results
 
-        hspec $ mapM_ (test runghc env' currDir origStackRoot newHome newStackRoot) tests
+  unless (null successes) $ do
+    logInfo "Successful tests:"
+    for_ successes $ \(x, _) -> logInfo $ "- " <> display x
+    logInfo ""
 
-hasTest :: FilePath -> FilePath -> IO Bool
-hasTest root dir = doesFileExist $ root </> dir </> "Main.hs"
+  if null failures
+    then logInfo "No failures!"
+    else do
+      logInfo "Failed tests:"
+      for_ failures $ \(x, ec) -> logInfo $ "- " <> display x <> " - " <> displayShow ec
+      liftIO exitFailure
 
-test :: FilePath -- ^ runghc
-     -> [(String, String)] -- ^ env
-     -> FilePath -- ^ currdir
-     -> FilePath -- ^ origStackRoot
-     -> FilePath -- ^ newHome
-     -> FilePath -- ^ newStackRoot
-     -> String
-     -> Spec
-test runghc env' currDir origStackRoot newHome newStackRoot name = it name $ withDir $ \dir -> do
-    newHomeExists <- doesDirectoryExist newHome
-    when newHomeExists (removeDirectoryRecursive newHome)
-    createDirectoryIfMissing True newStackRoot
-    copyTree toCopyRoot origStackRoot newStackRoot
-    writeFile (newStackRoot </> "config.yaml") "system-ghc: true\ninstall-ghc: false\n"
-    let testDir = currDir </> "tests" </> name
-        mainFile = testDir </> "Main.hs"
-        libDir = currDir </> "lib"
-        cp = (proc runghc
-                [ "-clear-package-db"
-                , "-global-package-db"
-                , "-i" ++ libDir
-                , mainFile
-                ])
-                { cwd = Just dir
-                , env = Just (Map.toList
-                            $ Map.insert "TEST_DIR" testDir
-                            $ Map.fromList env')
-                }
+data Speed = Fast | Slow
 
-    copyTree (const True) (testDir </> "files") dir
+exeExt :: String
+exeExt = if isWindows then ".exe" else ""
 
-    (ClosedStream, outSrc, errSrc, sph) <- streamingProcess cp
-    (out, err, ec) <- runConcurrently $ (,,)
-        <$> Concurrently (outSrc `connect` sinkLbs)
-        <*> Concurrently (errSrc `connect` sinkLbs)
-        <*> Concurrently (waitForStreamingProcess sph)
-    when (ec /= ExitSuccess) $ throwIO $ TestFailure out err ec
+isWindows :: Bool
+isWindows = os == "mingw32"
+
+runApp :: Speed -> RIO App a -> RIO SimpleApp a
+runApp speed inner = do
+  simpleApp <- ask
+  runghc <- findExecutable "runghc" >>= either throwIO pure
+  srcDir <- canonicalizePath ""
+  testsRoot <- canonicalizePath $ srcDir </> "test/integration"
+  libdir <- canonicalizePath $ testsRoot </> "lib"
+  myPath <- liftIO getExecutablePath
+
+  stack <- canonicalizePath $ takeDirectory myPath </> "stack" ++ exeExt
+  logInfo $ "Using stack located at " <> fromString stack
+  proc stack ["--version"] runProcess_
+
+  testDirs
+    <- runConduitRes
+     $ sourceDirectory (testsRoot </> "tests")
+    .| filterMC (liftIO . hasTest)
+    .| foldMapC Set.singleton
+
+  let modifyEnvCommon
+        = Map.insert "SRC_DIR" (fromString srcDir)
+        . Map.insert "STACK_EXE" (fromString stack)
+        . Map.delete "GHC_PACKAGE_PATH"
+        . Map.fromList
+        . map (first T.toUpper)
+        . Map.toList
+
+  case speed of
+    Fast -> do
+      let app = App
+            { appSimpleApp = simpleApp
+            , appRunghc = runghc
+            , appLibDir = libdir
+            , appSetupHome = pure ()
+            , appTestDirs = testDirs
+            }
+      runRIO app $ withModifyEnvVars modifyEnvCommon inner
+    Slow -> do
+      morigStackRoot <- liftIO $ lookupEnv "STACK_ROOT"
+      origStackRoot <-
+        case morigStackRoot of
+          Nothing -> getAppUserDataDirectory "stack"
+          Just x -> pure x
+
+      withSystemTempDirectory "stackhome" $ \newHome -> withSystemTempDirectory "stack" $ \newStackRoot -> do
+        logInfo "Initializing/updating the original Pantry store"
+        proc stack ["update"] runProcess_
+        copyTree (origStackRoot </> "pantry") (newStackRoot </> "pantry")
+        let modifyEnv
+                 = Map.insert "HOME" (fromString newHome)
+                 . Map.insert "APPDATA" (fromString newHome)
+                 . Map.insert "STACK_ROOT" (fromString newStackRoot)
+                 . modifyEnvCommon
+
+            app = App
+              { appSimpleApp = simpleApp
+              , appRunghc = runghc
+              , appLibDir = libdir
+              , appSetupHome = do
+                  newHomeExists <- doesDirectoryExist newHome
+                  when newHomeExists (removeDirectoryRecursive newHome)
+                  createDirectoryIfMissing True newHome
+
+                  createDirectoryIfMissing True newStackRoot
+                  runConduitRes $
+                    sourceDirectory newStackRoot .| mapM_C (\entry -> do
+                      let name = takeFileName entry
+                      unless (name == "." || name == ".." || name == "pantry") $ do
+                        isFile <- doesFileExist entry
+                        if isFile
+                          then removeFile entry
+                          else removeDirectoryRecursive entry
+                      )
+                  writeFileBinary (newStackRoot </> "config.yaml") "system-ghc: true\ninstall-ghc: false\n"
+              , appTestDirs = testDirs
+              }
+
+        runRIO app $ withModifyEnvVars modifyEnv inner
+
+
+hasTest :: FilePath -> IO Bool
+hasTest dir = doesFileExist $ dir </> "Main.hs"
+
+data App = App
+  { appRunghc :: !FilePath
+  , appLibDir :: !FilePath
+  , appSetupHome :: !(RIO App ())
+  , appSimpleApp :: !SimpleApp
+  , appTestDirs :: !(Set FilePath)
+  }
+simpleAppL :: Lens' App SimpleApp
+simpleAppL = lens appSimpleApp (\x y -> x { appSimpleApp = y })
+instance HasLogFunc App where
+  logFuncL = simpleAppL.logFuncL
+instance HasProcessContext App where
+  processContextL = simpleAppL.processContextL
+
+test :: FilePath -- ^ test dir
+     -> RIO App (Map Text ExitCode)
+test testDir = withDir $ \dir -> do
+    join $ asks appSetupHome
+
+    runghc <- asks appRunghc
+    libDir <- asks appLibDir
+    let mainFile = testDir </> "Main.hs"
+
+    copyTree (testDir </> "files") dir
+
+    withSystemTempFile (name <.> "log") $ \logfp logh -> do
+      ec <- withWorkingDir dir
+          $ withModifyEnvVars (Map.insert "TEST_DIR" $ fromString testDir)
+          $ proc runghc
+              [ "-clear-package-db"
+              , "-global-package-db"
+              , "-i" ++ libDir
+              , mainFile
+              ]
+           $ runProcess
+           . setStdin closed
+           . setStdout (useHandleOpen logh)
+           . setStderr (useHandleOpen logh)
+      hClose logh
+
+      case ec of
+        ExitSuccess -> logInfo "Success!"
+        _ -> do
+          logError "Failure, dumping log\n\n"
+          withSourceFile logfp $ \src ->
+            runConduit $ src .| stderrC
+          logError $ "\n\nEnd of log for " <> fromString name
+      pure $ Map.singleton (fromString name) ec
   where
+    name = takeFileName testDir
     withDir = withSystemTempDirectory ("stack-integration-" ++ name)
 
-data TestFailure = TestFailure L.ByteString L.ByteString ExitCode
-    deriving Typeable
-instance Show TestFailure where
-    show (TestFailure out err ec) = concat
-        [ "Exited with " ++ show ec
-        , "\n\nstdout:\n"
-        , toStr out
-        , "\n\nstderr:\n"
-        , toStr err
-        ]
-      where
-        toStr = TL.unpack . TL.decodeUtf8With lenientDecode
-instance Exception TestFailure
-
-copyTree :: (FilePath -> Bool) -> FilePath -> FilePath -> IO ()
-copyTree toCopy src dst =
-    runResourceT (sourceDirectoryDeep False src `connect` CL.mapM_ go)
+copyTree :: MonadIO m => FilePath -> FilePath -> m ()
+copyTree src dst =
+    liftIO $
+    runResourceT (sourceDirectoryDeep False src `connect` mapM_C go)
         `catch` \(_ :: IOException) -> return ()
   where
-    go srcfp = when (toCopy srcfp) $ liftIO $ do
+    go srcfp = liftIO $ do
         Just suffix <- return $ stripPrefix src srcfp
         let dstfp = dst </> stripHeadSeparator suffix
         createDirectoryIfMissing True $ takeDirectory dstfp
@@ -139,16 +228,3 @@ copyTree toCopy src dst =
     stripHeadSeparator fp@(x:xs) = if isPathSeparator x
                                    then xs
                                    else fp
-
-
-toCopyRoot :: FilePath -> Bool
-toCopyRoot srcfp = any (`isSuffixOf` srcfp)
-    -- FIXME command line parameters to control how many of these get
-    -- copied, trade-off of runtime/bandwidth vs isolation of tests
-    [ ".tar"
-    , ".tar.gz"
-    , ".xz"
-    -- , ".gz"
-    , ".7z.exe"
-    , "00-index.cache"
-    ]

--- a/test/integration/IntegrationSpec.hs
+++ b/test/integration/IntegrationSpec.hs
@@ -1,6 +1,7 @@
 {-# LANGUAGE BangPatterns        #-}
 {-# LANGUAGE NoImplicitPrelude   #-}
 {-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE RankNTypes          #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 
 import           Conduit
@@ -108,7 +109,7 @@ runApp speed inner = do
             { appSimpleApp = simpleApp
             , appRunghc = runghc
             , appLibDir = libdir
-            , appSetupHome = pure ()
+            , appSetupHome = id
             , appTestDirs = testDirs
             }
       runRIO app $ withModifyEnvVars modifyEnvCommon inner
@@ -119,40 +120,31 @@ runApp speed inner = do
           Nothing -> getAppUserDataDirectory "stack"
           Just x -> pure x
 
-      withSystemTempDirectory "stackhome" $ \newHome -> withSystemTempDirectory "stack" $ \newStackRoot -> do
-        logInfo "Initializing/updating the original Pantry store"
-        proc stack ["update"] runProcess_
-        copyTree (origStackRoot </> "pantry") (newStackRoot </> "pantry")
-        let modifyEnv
-                 = Map.insert "HOME" (fromString newHome)
-                 . Map.insert "APPDATA" (fromString newHome)
-                 . Map.insert "STACK_ROOT" (fromString newStackRoot)
-                 . modifyEnvCommon
+      logInfo "Initializing/updating the original Pantry store"
+      proc stack ["update"] runProcess_
 
-            app = App
-              { appSimpleApp = simpleApp
-              , appRunghc = runghc
-              , appLibDir = libdir
-              , appSetupHome = do
-                  newHomeExists <- doesDirectoryExist newHome
-                  when newHomeExists (removeDirectoryRecursive newHome)
-                  createDirectoryIfMissing True newHome
+      pantryRoot <- canonicalizePath $ origStackRoot </> "pantry"
+      let modifyEnv
+               = Map.insert "PANTRY_ROOT" (fromString pantryRoot)
+               . modifyEnvCommon
 
-                  createDirectoryIfMissing True newStackRoot
-                  runConduitRes $
-                    sourceDirectory newStackRoot .| mapM_C (\entry -> do
-                      let name = takeFileName entry
-                      unless (name == "." || name == ".." || name == "pantry") $ do
-                        isFile <- doesFileExist entry
-                        if isFile
-                          then removeFile entry
-                          else removeDirectoryRecursive entry
-                      )
-                  writeFileBinary (newStackRoot </> "config.yaml") "system-ghc: true\ninstall-ghc: false\n"
-              , appTestDirs = testDirs
-              }
+          app = App
+            { appSimpleApp = simpleApp
+            , appRunghc = runghc
+            , appLibDir = libdir
+            , appSetupHome = \inner' -> withSystemTempDirectory "home" $ \newHome -> do
+                let newStackRoot = newHome </> ".stack"
+                createDirectoryIfMissing True newStackRoot
+                let modifyEnv'
+                      = Map.insert "HOME" (fromString newHome)
+                      . Map.insert "APPDATA" (fromString newHome)
+                      . Map.insert "STACK_ROOT" (fromString newStackRoot)
+                writeFileBinary (newStackRoot </> "config.yaml") "system-ghc: true\ninstall-ghc: false\n"
+                withModifyEnvVars modifyEnv' inner'
+            , appTestDirs = testDirs
+            }
 
-        runRIO app $ withModifyEnvVars modifyEnv inner
+      runRIO app $ withModifyEnvVars modifyEnv inner
 
 
 hasTest :: FilePath -> IO Bool
@@ -161,7 +153,7 @@ hasTest dir = doesFileExist $ dir </> "Main.hs"
 data App = App
   { appRunghc :: !FilePath
   , appLibDir :: !FilePath
-  , appSetupHome :: !(RIO App ())
+  , appSetupHome :: !(forall a. RIO App a -> RIO App a)
   , appSimpleApp :: !SimpleApp
   , appTestDirs :: !(Set FilePath)
   }
@@ -172,11 +164,15 @@ instance HasLogFunc App where
 instance HasProcessContext App where
   processContextL = simpleAppL.processContextL
 
+-- | Call 'appSetupHome' on the inner action
+withHome :: RIO App a -> RIO App a
+withHome inner = do
+  app <- ask
+  appSetupHome app inner
+
 test :: FilePath -- ^ test dir
      -> RIO App (Map Text ExitCode)
-test testDir = withDir $ \dir -> do
-    join $ asks appSetupHome
-
+test testDir = withDir $ \dir -> withHome $ do
     runghc <- asks appRunghc
     libDir <- asks appLibDir
     let mainFile = testDir </> "Main.hs"

--- a/test/integration/tests/2433-ghc-by-version/files/run.sh
+++ b/test/integration/tests/2433-ghc-by-version/files/run.sh
@@ -2,10 +2,10 @@
 
 set -exuo pipefail
 
-export PATH=$(pwd)/fake-path:$($STACK_EXE path --resolver ghc-8.2.2 --compiler-bin):$PATH
+export PATH=$(pwd)/fake-path:$("$STACK_EXE" path --resolver ghc-8.2.2 --compiler-bin):$PATH
 export STACK_ROOT=$(pwd)/fake-root
 
 which ghc
 
-$STACK_EXE --system-ghc --no-install-ghc --resolver ghc-8.2.2 ghc -- --info
-$STACK_EXE --system-ghc --no-install-ghc --resolver ghc-8.2.2 runghc foo.hs
+"$STACK_EXE" --system-ghc --no-install-ghc --resolver ghc-8.2.2 ghc -- --info
+"$STACK_EXE" --system-ghc --no-install-ghc --resolver ghc-8.2.2 runghc foo.hs

